cli/command/telemetry_utils.go

```diff
@@ -0,0 +1,172 @@
+//
+// can also be used for spans!
+func InstrumentCobraCommands(ctx context.Context, cmd *cobra.Command, meter metric.Meter) {
+	// If PersistentPreRunE is nil, make it execute PersistentPreRun and return nil by default
+	ogPersistentPreRunE := cmd.PersistentPreRunE
+	if ogPersistentPreRunE == nil {
+		ogPersistentPreRun := cmd.PersistentPreRun
+		ogPersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+			ogPersistentPreRun(cmd, args)
+			return nil
+		}
+		cmd.PersistentPreRun = nil
+	}
+
+	// wrap RunE in PersistentPreRunE so that this operation gets executed on all children commands
+	cmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+		// If RunE is nil, make it execute Run and return nil by default
+		ogRunE := cmd.RunE
+		if ogRunE == nil {
+			ogRun := cmd.Run
+			ogRunE = func(cmd *cobra.Command, args []string) error {
+				ogRun(cmd, args)
+				return nil
+			}
+			cmd.Run = nil
+		}
+		cmd.RunE = func(cmd *cobra.Command, args []string) error {
+			// insert the meter and tracer objects into the context
+			// so commands can use them in their RunE func
+			newCtx := context.WithValue(cmd.Context(), OtelMeterCtxKey, meter)
+			// newCtx := context.WithValue(newCtx, OtelTracerCtxKey, tracer))
+			cmd.SetContext(newCtx)
+
+			// start the timer as the first step of every cobra command
+			stopCobraCmdTimer := startCobraCommandTimer(cmd)
+			cmdErr := ogRunE(cmd, args)
+			stopCobraCmdTimer(cmdErr)
+			return cmdErr
+		}
+
+		return ogPersistentPreRunE(cmd, args)
+	}
+}
+
+func startCobraCommandTimer(cmd *cobra.Command) func(err error) {
+	ctx := cmd.Context()
+	meter, ok := ctx.Value(OtelMeterCtxKey).(metric.Meter)
+	if meter == nil || !ok {
+		otel.Handle(errors.New("meter not found in ctx."))
+		// return a no-op if the meter cannot be found in the command's ctx
+		return func(err error) {}
+	}
+	baseAttrs := BaseMetricAttributes(cmd)
+	durationCounter, _ := meter.Float64Counter(
+		"command.duration",
+		metric.WithDescription("Measures the duration of the cobra command"),
+		metric.WithUnit("ms"),
+	)
+	start := time.Now()
+
+	return func(err error) {
+		duration := float64(time.Since(start)) / float64(time.Millisecond)
+		extraAttrs := []attribute.KeyValue{}
+		exitCode := 0
+		if err != nil {
+			exitCode = 1
+			if stderr, ok := err.(statusError); ok {
+				// StatusError should only be used for errors, and all errors should
+				// have a non-zero exit status, so only set this here if this value isn't 0
+				if stderr.StatusCode != 0 {
+					exitCode = stderr.StatusCode
+				}
+			}
+			// TODO(krissetto): this attribute might not be needed, but buildx has it. Check if it's necessary
+			extraAttrs = append(extraAttrs, attribute.String("command.error.type", otelErrorType(err)))
+		}
+		extraAttrs = append(extraAttrs, attribute.String("command.status", mapExitCode(exitCode)))
+		extraAttrs = append(extraAttrs, attribute.String("command.exit_code", strconv.Itoa(exitCode)))
```

Comment:
I'd recommend consolidating these into a single value for now and doing something like naming it `command.status.code`. With this more explicit namespacing, it's easy to add a `command.status.text` or something else in the future if we need it. The reason why I say to only use one is it seems like the map exit code can be done more easily and appropriately in the otel collector than here. If you push more logic into the otel collector, it's fewer things to maintain here and less disruptive if we decide to arbitrarily change the text.

Reply:
Sounds good to me